File: common/src/test/java/com/terrasect/common/testing/SnapshotOutputPaths.java
package com.terrasect.common.testing;

import java.io.File;
import java.nio.file.Path;

public final class SnapshotOutputPaths {
    private static final String COLLAPSED_PREFIX = "com.terrasect.common";

    public static File forTestClass(Class<?> testClass, String... segments) {
        var base = Path.of("build", "test-snapshots");
        var path = base.resolve(collapsedClassPath(testClass));
        if (segments != null) {
            for (String segment : segments) {
                if (segment == null || segment.isEmpty()) continue;
                path = path.resolve(segment);
            }
        }
        return path.toFile();
    }

    private static Path collapsedClassPath(Class<?> testClass) {
        var packageName = testClass.getPackageName();
        if (packageName.startsWith(COLLAPSED_PREFIX)) {
            packageName = packageName.substring(COLLAPSED_PREFIX.length());
            if (packageName.startsWith(".")) {
                packageName = packageName.substring(1);
            }
        }
        var path = packageName.isEmpty()
                ? Path.of(testClass.getSimpleName())
                : Path.of(packageName.replace('.', File.separatorChar)).resolve(testClass.getSimpleName());
        return path;
    }

    private SnapshotOutputPaths() {
    }
}

File: common/bin/test/terrasect/testing/SnapshotOutputPaths.kt
package terrasect.testing

import java.awt.image.BufferedImage
import java.io.File
import java.nio.file.Path
import javax.imageio.ImageIO
import org.junit.jupiter.api.Assertions.assertTrue

object SnapshotOutputPaths {
  private const val COLLAPSED_PREFIX = "terrasect"

  @JvmStatic
  fun forTestClass(testClass: Class<*>, vararg segments: String?): File {
    var path = Path.of("build", "test-snapshots").resolve(collapsedClassPath(testClass))
    for (segment in segments) {
      if (segment.isNullOrEmpty()) {
        continue
      }
      path = path.resolve(segment)
    }
    return path.toFile()
  }

  private fun collapsedClassPath(testClass: Class<*>): Path {
    var packageName = testClass.packageName
    if (packageName.startsWith(COLLAPSED_PREFIX)) {
      packageName = packageName.substring(COLLAPSED_PREFIX.length)
      if (packageName.startsWith(".")) {
        packageName = packageName.substring(1)
      }
    }
    return if (packageName.isEmpty()) {
      Path.of(testClass.simpleName)
    } else {
      Path.of(packageName.replace('.', File.separatorChar)).resolve(testClass.simpleName)
    }
  }
}

fun writeSnapshotPng(testClass: Class<*>, snapshotName: String, image: BufferedImage) {
  val outputFile = SnapshotOutputPaths.forTestClass(testClass, snapshotName)
  outputFile.parentFile.mkdirs()
  val written = ImageIO.write(image, "png", outputFile)
  assertTrue(written, "Expected to write PNG snapshot to ${outputFile.absolutePath}")
}
